barretenberg/cpp/src/barretenberg/dsl/acir_format/avm2_recursion_constraint.cpp

```diff
@@ -0,0 +1,240 @@
+ */
+aggregation_state_ct create_avm2_recursion_constraints(Builder& builder,
+                                                       const RecursionConstraint& input,
+                                                       const aggregation_state_ct& input_aggregation_object,
+                                                       bool has_valid_witness_assignments)
+{
+    using Flavor = avm2::AvmRecursiveFlavor_<Builder>;
+    using RecursiveVerificationKey = Flavor::VerificationKey;
+    using RecursiveVerifier = avm2::AvmRecursiveVerifier_<Flavor>;
+
+    ASSERT(input.proof_type == AVM);
+
+    auto fields_from_witnesses = [&](std::vector<uint32_t> const& input) {
+        std::vector<field_ct> result;
+        result.reserve(input.size());
+        for (const auto& idx : input) {
+            result.emplace_back(field_ct::from_witness_index(&builder, idx));
+        }
+        return result;
+    };
+
+    // Construct in-circuit representation of the verification key, proof and public inputs
+    const auto key_fields = fields_from_witnesses(input.key);
+    const auto proof_fields = fields_from_witnesses(input.proof);
+    // We assume a single public input value for now (reverted flag)
+    const auto public_inputs_flattened = fields_from_witnesses(input.public_inputs);
+
+    // Populate the key fields and proof fields with dummy values to prevent issues (e.g. points must be on curve).
+    if (!has_valid_witness_assignments) {
+        create_dummy_vkey_and_proof(builder, input.proof.size(), input.public_inputs.size(), key_fields, proof_fields);
+    }
+
+    // Recursively verify the proof
+    auto vkey = std::make_shared<RecursiveVerificationKey>(builder, key_fields);
+    RecursiveVerifier verifier(builder, vkey);
+
+    // TODO: We currently pass public_inputs_flattened as a single column.
+    //       Once we deploy multiple public inputs columns, we will have to transform the
+    //       public_inputs_flattened vector to a vector of vectors.
+    aggregation_state_ct output_agg_object =
+        verifier.verify_proof(proof_fields, { public_inputs_flattened }, input_aggregation_object);
+
+    return output_agg_object;
+}
+
+/**
+ * @brief Add constraints associated with recursive verification of an AVM2 proof using Goblin
+ *
+ * @param builder
+ * @param input
+ * @param input_aggregation_object_indices
+ * @param has_valid_witness_assignments
+ * @return HonkRecursionConstraintOutput {pairing agg object, ipa claim, ipa proof}
+ */
+HonkRecursionConstraintOutput create_avm2_recursion_constraints_goblin(
+    Builder& builder,
+    const RecursionConstraint& input,
+    const aggregation_state_ct& input_aggregation_object,
+    bool has_valid_witness_assignments)
+{
+    using RecursiveVerifier = avm2::AvmGoblinRecursiveVerifier;
+
+    ASSERT(input.proof_type == AVM);
+
+    auto fields_from_witnesses = [&](std::vector<uint32_t> const& input) {
```

Comment:
if you need to change anything else, please change these to `const std::vector<uint32_t>& input` which is what we are using now